opal/mca/common/ucx/common_ucx_wpool.c

```diff
@@ -447,12 +448,14 @@ int opal_common_ucx_wpmem_create(opal_common_ucx_ctx_t *ctx, void **mem_base, si
         goto error_rkey_pack;
     }
 
-    ret = exchange_func(rkey_addr, rkey_addr_len, &mem->mem_addrs, &mem->mem_displs,
-                        exchange_metadata);
-    if (ret != OPAL_SUCCESS) {
-        goto error_rkey_pack;
+    if (exchange_mode == OPAL_COMMON_UCX_WPMEM_ADDR_EXCHANGE_FULL) {
```

Comment:
where's the place where you cover the OPAL_COMMON_UCX_WPMEM_ADDR_EXCHANGE_DIRECT case?

Reply:
During get_dynamic_win_info, when we evaluate the temp_dynamic_buffer, the mem->addr will be NULL, because it wasn't set during win_attach (with DIRECT). Then we copy over the addresses that we got from the dynamic win.

Reply:
This is also very inneficcient during rget rput, because we get the dynamic win twice. and we get sevel kb, it's really bad, let's merge the fix, and we can work on a better solution.